tensorflow/java/src/main/java/org/tensorflow/Server.java

```diff
@@ -0,0 +1,134 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+package org.tensorflow;
+
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+/**
+ * An in-process TensorFlow server, for use in distributed training.
+ *
+ * A {@code Server} instance encapsulates a set of devices and a
+ * {@link org.tensorflow.Session} target that can participate in distributed
+ * training. A server belongs to a cluster (specified by a
+ * {@code ClusterSpec}), and corresponds to a particular task in a named job.
+ * The server can communicate with any other server in the same cluster.
+ *
+ * <p><b>WARNING:</b> A {@code Server} owns resources that <b>must</b> be
+ * explicitly freed by invoking {@link #close()}.
+ *
+ * <p>Instances of a {@code Server} are thread-safe.
+ *
+ * <p>Using example:
+ * <pre>
+ * {@code
+ * ClusterDef clusterDef = ClusterDef.newBuilder()
+ *   .addJob(JobDef.newBuilder()
+ *   .setName("worker")
+ *   .putTasks(0, "localhost:4321")
+ *   .build()
+ * ).build();
+ *
+ * ServerDef serverDef = ServerDef.newBuilder()
+ *   .setCluster(clusterDef)
+ *   .setJobName("worker")
+ *   .setTaskIndex(0)
+ *   .setProtocol("grpc")
+ * .build();
+ *
+ * try (Server srv = new Server(serverDef.toByteArray())) {
+ *   srv.start();
+ *   srv.join();
+ * }
+ * }
+ * </pre>
+ */
+public final class Server implements AutoCloseable {
+
+  /** 
+   * Constructs a new instance of server. 
+   *
+   * @param serverDef Server definition specified as a serialized
+   *        <a href="https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/tensorflow_server.proto">ServerDef</a>
+   *        protocol buffer.
+   */
+  public Server(byte[] serverDef) {
+    nativeHandle = allocate(serverDef);
+  }
+
+  /** Starts this server. */
+  public void start() {
+    lock.readLock().lock();
+    try {
+      start(nativeHandle);
+    }
+    finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  /** Stops this server. */
+  public void stop() {
+    lock.readLock().lock();
+    try {
+      stop(nativeHandle);
+    }
+    finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  /** Blocks until the server has shut down (currently blocks forever). */
+  public void join() {
+    lock.readLock().lock();
```

Comment:
Do you think it's better than a simple `RWLock`? I see at least one issue in you example (`notifyAll` should be called instead of `notify`) and two tricky parts (`numJoining` and `wait`/`while`). `RWLock` looks much easier.

Reply:
Yes should be `notifyAll`. 
What is the danger with the `while` and the `numJoining` counter?

I find the `RWLock` harder to follow, particularly since it is easy to gloss over the subtlety that `join` is a "write" operation.

Isn't this scenario exactly what the object monitor and `wait`/notify mechanism is meant for?

Reply:
The `numJoining` is danger because it's quite hard to understand if it should be `volatile` in this case or not. Method `wait()` it tricky because it should be wrapped into `while` loop.

Well, anyway, I've been thinking about locks because the problems for me looks like "forbid parallel execution of `close()` and other methods". Anyway, if you think `synchronized/wait/notify` approach is better I'll update the code.

Reply:
Done.